Updated __repr__ for Port class

summeraz opened the pull request:
Since `Port` inherits from `Compound` the print representation was the same as the `Compound` class and was a bit confusing, e.g.
```
<Port 0 particles, non-periodic, 0 bonds, id: 4495913424>
```
Not very helpful since this will always show 0 particles, non-periodic, and 0 bonds.

The updated representation now looks like this:
```
<Port, direction: (0.2673, 0.5345, 0.8018), position: (0.0267, 0.0535, 0.0802), anchor: <C pos=( 0.0000, 0.0000, 0.0000), 0 bonds, id: 4495687240>, id: 4496003648>
```
where the `Port`'s `direction`, `position`, and `anchor` are shown.

I also think it would be useful to print the label that has been provided to a `Port` if it has been added to the hierarchy of another mBuild `Compound`. For example, if a `Port` was added to a `Compound` and provided with the label `'port'` the representation would look like this:
```
<Port, direction: (0.2673, 0.5345, 0.8018), position: (0.0267, 0.0535, 0.0802), anchor: <C pos=( 0.0000, 0.0000, 0.0000), 0 bonds, id: 4495687240>, Labels: ['port'], id: 4496003648>
```
There may be a better way to provide this information however.

justinGilmer commented:
Perhaps instead of `Labels: ...`, `to access: compoundVariableName['up']`

Not sure how easy that would be. esp if the hierarchy is complex. What compound would be used in the hierarchy? 

But this would be nice to have.

ctk3b commented:
Definitely like the idea of refining this. +1 to adding info about the anchor and the label but I think with the direction this is now becoming a bit too long. In general the use case for the repr is when printing to something like stdout and I think this will clutter the screen a bit too much.

ctk3b commented:
Any further thoughts on this?

summeraz commented:
Sorry I had kind of forgotten about this. I agree, the repr is too long with all of the information that I've added. I think `anchor` is definitely useful, although that output is actually pretty long (since `anchor` is basically the repr for that `Compound`). Maybe we just provide the `Compound` name instead of the full repr?

I think we can probably ditch `position` to help make this more succinct. I do kind of like having `direction` in the repr, although `Labels` is certainly more useful information.

As far as `Labels` is concerned, in the code I've modified thus far I'm only looking for labels in `parent.labels`. A better way would probably be to look in the labels for each `Compound` in `Port.referrers`.

I'll work on some of these changes and will try to get some updated code committed later tonight.

summeraz commented:
Okay I've made some changes. I've added an additional property for `Ports` that I call `access_labels`. These are the labels (relative to the `self.root`) that can be used to access the `Port`. This seems like something that would be useful, although I'm not sure the object (a list of strings) that this property returns is the most useful. Maybe just having this return a dictionary of the form {referrer: label} would be cleaner?

I've also shortened `__repr__` to include only the `Port`'s name, the `anchor` name, the labels from `access_labels`, and the unique identifier.

ctk3b commented on mbuild/port.py:
```diff
@@ -84,3 +86,40 @@ def direction(self):
+                               if val == referrer]
+            port_labels = [key for key, val in referrer.labels.items()
+                           if val == self]
+            if referrer == self.root:
```
I think you want this comparison to be done via identity: `referrer is self.root`
